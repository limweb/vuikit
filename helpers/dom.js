const boundEvents = []

// add event listener shorthand
export function on (el, event, handler, namespace = 'def') {
  boundEvents[namespace] = boundEvents[namespace] || []
  boundEvents[namespace].push({ el, event, handler })
  el.addEventListener(event, handler)
}

export function offAll (namespace = 'def') {
  if (boundEvents[namespace] !== undefined) {
    for (let i = 0; i < boundEvents[namespace].length; ++i) {
      const { el, event, handler } = boundEvents[namespace][i]
      el.removeEventListener(event, handler)
    }
  }
}

export function addClass (el, className) {
  if (el.classList) {
    el.classList.add(className)
  } else {
    // IE9
    el.className += ' ' + className
  }
}

export function hasClass (el, className) {
  return el.classList.contains(className)
}

export function removeClass (el, className) {
  if (el.classList) {
    el.classList.remove(className)
  } else {
    // IE9
    el.className = el.className.replace(new RegExp(`^${className}$`), '')
  }
}

export function css (el, style) {
  return window.getComputedStyle(el)[style]
}

const UA = window.navigator.userAgent.toLowerCase()
const isIE9 = UA && UA.indexOf('msie 9.0') > 0

// Transition property/event sniffing
export let transitionProp = 'transition'
export let transitionEndEvent = 'transitionend'
export let animationProp = 'animation'
export let animationEndEvent = 'animationend'

if (!isIE9) {
  /* istanbul ignore if */
  if (window.ontransitionend === undefined &&
    window.onwebkittransitionend !== undefined) {
    transitionProp = 'WebkitTransition'
    transitionEndEvent = 'webkitTransitionEnd'
  }
  if (window.onanimationend === undefined &&
    window.onwebkitanimationend !== undefined) {
    animationProp = 'WebkitAnimation'
    animationEndEvent = 'webkitAnimationEnd'
  }
}

/**
 * Get the closest matching element up the DOM tree.
 * @private
 * @param  {Element} elem     Starting element
 * @param  {String}  selector Selector to match against
 * @return {Boolean|Element}  Returns null if not match found
 */
export function getClosest (elem, selector) {
  // Element.matches() polyfill
  const Element = window.Element
  if (!Element.prototype.matches) {
    Element.prototype.matches =
      Element.prototype.matchesSelector ||
      Element.prototype.mozMatchesSelector ||
      Element.prototype.msMatchesSelector ||
      Element.prototype.oMatchesSelector ||
      Element.prototype.webkitMatchesSelector ||
      function (s) {
        const matches = (this.document || this.ownerDocument).querySelectorAll(s)
        let i = matches.length
        while (--i >= 0 && matches.item(i) !== this) {}
        return i > -1
      }
  }

  // Get closest match
  for (; elem && elem !== document; elem = elem.parentNode) {
    if (elem.matches(selector)) return elem
  }

  return null
}
