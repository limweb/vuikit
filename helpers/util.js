/**
 * Utility functions
 */

/**
 * Quick object check - this is primarily used to tell
 * Objects from primitive values when we know the value
 * is a JSON-compliant type.
 */
export function isObject (obj) {
  return obj !== null && typeof obj === 'object'
}

/**
 * Strict object type check. Only returns true
 * for plain JavaScript objects.
 */
const toString = Object.prototype.toString
const OBJECT_STRING = '[object Object]'
export function isPlainObject (obj) {
  return toString.call(obj) === OBJECT_STRING
}

/**
 * Check if value is plain string
 */
export function isString (val) {
  return typeof val === 'string'
}

/**
 * Check if value is plain integer
 */
export function isInteger (val) {
  return Number.isInteger(val)
}

/**
 * Check if value is plain function
 */
export function isFunction (val) {
  return typeof val === 'function'
}

/**
 * Check if value is plain array
 */
export const isArray = Array.isArray

/**
 * Check if value is in array
 */
export function inArray (array, value) {
  return (array || []).indexOf(value) !== -1
}

/**
* Convert an Array-like object or value to Array
*/
export function toArray (value) {
  if (isArray(value)) {
    return value
  }
  return isObject(value)
    ? _toArray(value)
    : [value]
}

function _toArray (list, start) {
  start = start || 0
  let i = list.length - start
  const ret = new Array(i)
  while (i--) {
    ret[i] = list[i + start]
  }
  return ret
}

/**
 * Returns an array range
 */
export function range (start, stop, step = 1) {
  if (typeof stop === 'undefined') {
    stop = start
    start = 0
  }
  return Array.from(new Array(Math.floor((stop - start) / step)), (x, i) => start + i * step)
}

export function each (obj, iterator) {
  var i, key
  if (typeof obj.length === 'number') {
    for (i = 0; i < obj.length; i++) {
      iterator.call(obj[i], obj[i], i)
    }
  } else if (isObject(obj)) {
    for (key in obj) {
      if (obj.hasOwnProperty(key)) {
        iterator.call(obj[key], obj[key], key)
      }
    }
  }
  return obj
}

export function merge (target) {
  const args = Array.prototype.slice.call(arguments, 1)
  args.forEach((source) => {
    _merge(target, source, true)
  })
  return target
}

function _merge (target, source, deep) {
  for (var key in source) {
    if (deep && (isPlainObject(source[key]) || isArray(source[key]))) {
      if (isPlainObject(source[key]) && !isPlainObject(target[key])) {
        target[key] = {}
      }
      if (isArray(source[key]) && !isArray(target[key])) {
        target[key] = []
      }
      _merge(target[key], source[key], deep)
    } else if (source[key] !== undefined) {
      target[key] = source[key]
    }
  }
}

/**
 * Warn about errors only in no production
 */

let warn

if (process.env.NODE_ENV !== 'production') {
  const hasConsole = typeof console !== 'undefined'
  warn = (name, msg) => {
    if (hasConsole) {
      console.error(msg === undefined
        ? `[Vuikit]: ${msg}`
        : `[${name} warn]: ${msg}`
      )
    }
  }
}

export { warn }
