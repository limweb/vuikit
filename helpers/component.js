import { each, merge } from './util'

export function filterByTag (nodes, tag) {
  const result = []
  each(nodes, node => {
    if (node.componentOptions && node.componentOptions.tag === tag) {
      result.push(node)
    }
  })
  return result
}

export function getProps (vm) {
  return vm.componentOptions.propsData
}

export function getFinalProps (component) {
  let props = {}
  while (component) {
    props = merge(props, component.props)
    component = component.extends
  }
  return props
}
