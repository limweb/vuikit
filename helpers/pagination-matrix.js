import { range as _range } from './util'

/**
 * Returns an array with represented ranges pages
 */
export default function ({ active = 1, total = 200, limit = 10, range = 3 }) {
  const matrix = []
  const totalPages = Math.ceil(total / limit)
  // get main pages
  const mainPages = getMainPages({ active, range, totalPages })
  const first = mainPages[0]
  const last = mainPages[mainPages.length - 1]
  // get pre/post pages
  const prePages = _range(1, (first <= 3) ? first : 2)
  const postPages = _range(
    last >= (totalPages - 2) ? last + 1 : totalPages,
    totalPages + 1
  )

  let nextPage = 1
  ;[].concat(prePages, mainPages, postPages).forEach(page => {
    if (page === nextPage) {
      matrix.push(page)
      nextPage++
    } else {
      matrix.push('...')
      matrix.push(page)
      nextPage = page + 1
    }
  })

  return matrix
}

const getMainPages = ({ active, range, totalPages }) => {
  let start = active - range
  let end = active + range
  if (end > totalPages) {
    end = totalPages
    start = totalPages - range * 2
    start = start < 1 ? 1 : start
  }
  if (start <= 1) {
    start = 1
    end = Math.min(range * 2 + 1, totalPages)
  }
  return _range(start, end + 1)
}
