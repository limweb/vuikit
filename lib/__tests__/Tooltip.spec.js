import { createVue, destroyVM, triggerEvent, queryByTag, toJSON } from './_util'
import { each } from 'helpers/util'
import waitForUpdate from './_wait-for-update'

describe('Tooltip', () => {
  let vm
  afterEach(() => {
    destroyVM(vm)
  })

  it('renders correctly', () => {
    vm = createVue(`
      <button><vk-tooltip content="Content"></button>
    `)
    expect(toJSON(vm)).toMatchSnapshot()
  })

  it('fallback to parent target by default', done => {
    vm = createVue({
      template: `
        <div>
          <vk-button />
          <vk-tooltip :target="target" />
        </div>`,
      data: () => ({
        target: false
      })
    })

    const tooltip = queryByTag(vm, 'vk-tooltip')
    const button = queryByTag(vm, 'vk-tooltip')
    expect(tooltip.targetElement).toEqual(vm.$el)
    vm.target = button
    waitForUpdate(() => {
      expect(tooltip.targetElement).toEqual(button)
    }).then(done)
  })

  it('calculates the placement correctly', () => {
    vm = createVue(`
      <button><vk-tooltip placement="top" /></button>
    `)
    const tooltip = queryByTag(vm, 'vk-tooltip')
    expect(tooltip.$el.getAttribute('x-placement')).toEqual('top')
    expect(tooltip.$el.classList.contains('uk-tooltip-top')).toBeTruthy()

    // check placement conversion popper > uikit
    const placements = {
      'top': 'top',
      'top-start': 'top-left',
      'top-end': 'top-right',
      'right': 'right',
      'right-start': 'right',
      'right-end': 'right',
      'bottom': 'bottom',
      'bottom-start': 'bottom-left',
      'bottom-end': 'bottom-right',
      'left': 'left',
      'left-start': 'left',
      'left-end': 'left'
    }
    each(placements, (value, key) => {
      expect(tooltip.convertPlacement(key)).toEqual(value)
    })
  })

  it('displays when the show prop is set to true', done => {
    vm = createVue(`
      <button><vk-tooltip /></button>
    `)
    const tooltip = queryByTag(vm, 'vk-tooltip')
    const cb = jest.fn()

    tooltip.$on('show', cb)
    triggerEvent(vm.$el, 'mouseenter')
    waitForUpdate(() => {
      expect(tooltip.$el.style.display).not.toEqual('none')
      expect(cb).toHaveBeenCalled()
    }).then(done)
  })

  it('hides and triggers hide event on target mouseleave', done => {
    vm = createVue(`
      <button><vk-tooltip /></button>
    `)
    const tooltip = queryByTag(vm, 'vk-tooltip')
    const cb = jest.fn()

    tooltip.$on('hide', cb)
    triggerEvent(vm.$el, 'mouseenter')
    triggerEvent(vm.$el, 'mouseleave')
    waitForUpdate(() => {
      expect(tooltip.$el.style.display).toEqual('none')
      expect(cb).toHaveBeenCalled()
    }).then(done)
  })

  it('hides when show prop is set to false', done => {
    vm = createVue(`
      <button><vk-tooltip /></button>
    `)
    const tooltip = queryByTag(vm, 'vk-tooltip')
    const cb = jest.fn()

    tooltip.$on('hide', cb)
    triggerEvent(vm.$el, 'mouseenter')
    triggerEvent(vm.$el, 'mouseleave')
    waitForUpdate(() => {
      expect(tooltip.$el.style.display).toEqual('none')
      expect(cb).toHaveBeenCalled()
    }).then(done)
  })
})
