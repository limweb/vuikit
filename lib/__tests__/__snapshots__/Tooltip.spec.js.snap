exports[`Tooltip renders correctly 1`] = `
Object {
  "child": Array [
    Object {
      "node": "element",
      "tag": "button",
    },
  ],
  "node": "root",
}
`;
